src/api/fuel.ts

```diff
@@ -0,0 +1,535 @@
+
+// expire time in millisec
+const expireSeconds = 3600;
+
+const chain = getChain();
+const client = new APIClient({
+  url: chain.getHyperionEndpoint()
+});
+
+const fuelrpc = chain.getFuelRPCEndpoint();
+const resourceProviderEndpoint = `${fuelrpc.protocol}://${fuelrpc.host}:${fuelrpc.port}/v1/resource_provider/request_transaction`;
+
+// Auxiliar interfaces
+interface ResourceProviderResponse {
+  code: number;
+  data: ResponseData;
+}
+
+interface ResponseData {
+  request: [string, SignedTransaction];
+  signatures: Signature[];
+  version: number;
+  costs?: { ram: string };
+}
+
+interface SignedTransactionResponse extends SignTransactionResponse {
+  transaction: { signatures: Signature[] };
+}
+
+// Wrapper for the user to intersect the signTransaction call
+export class FuelUserWrapper extends User {
+  constructor(private user: User) {
+    super();
+  }
+
+  async signTransaction(
+    originalTransaction: AnyTransaction,
+    originalconfig?: SignTransactionConfig
+  ): Promise<SignTransactionResponse> {
+    try {
+      // Retrieve transaction headers
+      const info = await client.v1.chain.get_info();
+      const header = info.getTransactionHeader(expireSeconds);
+
+      // collect all contract abis
+      const abi_promises = originalTransaction.actions.map((a) =>
+        client.v1.chain.get_abi(a.account)
+      );
+      const responses = await Promise.all(abi_promises);
+      const abis = responses.map((x) => x.abi);
+      const abis_and_names = originalTransaction.actions.map((x, i) => ({
+        contract: x.account,
+        abi: abis[i]
+      }));
+
+      // create complete well formed transaction
+      const transaction = Transaction.from(
+        {
+          ...header,
+          actions: originalTransaction.actions
+        },
+        abis_and_names
+      );
+
+      // Pack the transaction for transport
+      const packedTransaction = PackedTransaction.from({
+        signatures: [],
+        packed_context_free_data: '',
+        packed_trx: Serializer.encode({ object: transaction })
+      });
+
+      const signer = PermissionLevel.from({
+        actor: (await this.user.getAccountName()) as NameType,
+        permission: this.requestPermission
+      });
+
+      // Submit the transaction to the resource provider endpoint
+      const cosigned = await fetch(resourceProviderEndpoint, {
+        body: JSON.stringify({
+          signer,
+          packedTransaction
+        }),
+        method: 'POST'
+      });
+
+      // Interpret the resulting JSON
+      const rpResponse =
+        (await cosigned.json()) as unknown as ResourceProviderResponse;
+
+      switch (rpResponse.code) {
+        case 402: {
+          // Resource Provider provided signature in exchange for a fee
+          // is ok to treat them with the same logic of code = 200?
+          // Yes acording to this: https://gist.github.com/aaroncox/d74a73b3d9fbc20836c32ea9deda5d70#file-fuel-core-presign-js-L128-L159
```

Comment:
What we do in the UI is prompt the user to accept the fee, if a fee exists on the transaction. Anchor itself also prompts for a fee - but since you're also integrating this into other wallets, it may be nice to prompt users to let them know the transaction includes a fee. 

![feebasedtx](https://user-images.githubusercontent.com/677686/208698007-146e5ddb-294e-4da2-a529-c0879bef3ee8.gif)

Reply:
@aaroncox we would still like to inform the user that this is free of charge, while encouraging him/her to power up the account for other dAPPs they might want to use going forward. As you mention, we also would like to cover other wallets too.

@Viterbo can you please verify the use case where fees are applied? Many thanks.